File: 3makingfriends/makingfriends.py
import sys
from collections import deque
from collections import defaultdict

LINE = sys.stdin.readline().split(' ')
N: int = int(LINE[0])
M: int = int(LINE[1])
parent = [i for i in range(0,N+1)]
# rank = [0]*(N+1)

def main():
    edges = buildEdges()
    #print(edges)
    print(kruskal(edges))


def kruskal(edges):
    sum = 0

    while(len(edges) != 0):
        u,v,w = edges.popleft()
        parentU = find(u)
        parentV = find(v)
        # rankU = rank[u]
        # rankV = rank[v]

        if parentU != parentV:
            # if rankU > rankV:
            #     parent[parentV] = parentU
            # elif rankU < rankV:
            #     parent[parentU] = parentV
            # else:
            #     parent[parentV] = parentU
            #     rank[u] += 1
            parent[parentV] = parentU
            sum += w

    return sum

def find(node):
    if parent[node] != node:
        parent[node] = find(parent[node])
    return parent[node]

def buildEdges():
    edges = []
    for _ in range(M):
        edges.append(createEdge())
    
    return deque(sorted(edges, key=lambda edge: edge[2]))

def createEdge():
    #edgeTuple = lambda lst: ((lst[0], lst[1]), lst[2]) # [x,y,z] -> ((x,y),z)
    return list(map(int, sys.stdin.readline().split(' ')))

if __name__ == '__main__':
    main()

File: 3makingfriends/makingfriends.hs
module Makingfriends where 
import Data.List (sortBy)
import Data.Function (on)
import Data.Sequence hiding (sortBy)

main = do 
    line <- getLine
    contents <- getContents
    let [n,_] = map read $ words line :: [Int]
        parent = [0..(n)]
        edges = sortBy (compare `on` (!!2)) $ buildEdges (lines contents)
    --print parent
    print edges
    --putStrLn "Running Kruskal..."
    result <- kruskal 0 edges (fromList parent)
    print result


kruskal :: Int -> [Edge] -> Seq Int -> IO (Int)
kruskal i   []      parent   = do 
    --print i
    return (i)
kruskal i ([u,v,w]:es) parent 
    | parentU /= parentV = do 
        --print (1,[u,v,w],debug)
        kruskal (i+w) es (union)
    | otherwise = do 
        --print (2,[u,v,w],debug)
        kruskal i es parent
    where 
        debug = (parent, i)
        (parentU,parent') = find u parent
        (parentV,newParent) = find v parent'
        union   = update parentV parentU newParent


type Edge = [Int]

buildEdges :: [String] -> [Edge]
buildEdges [] = []
buildEdges (l:ls) = edge l : buildEdges ls
    where 
        edge l = map read $ words l

find :: Int -> Seq Int -> (Int, Seq Int)
find n parent 
    | index parent n == n = (n, parent)
    | otherwise = (index (snd nextParent) p, update n (fst nextParent) (snd nextParent))
    where
        p = index parent n
        (nextParent) = find p parent

-- find :: [Int] -> [Int] -> Seq Int -> ([Int], Seq Int)
-- find [] res parent = (res,parent)
-- find (n:ns) res parent
--     | p == n = find (ns) (n:res) parent
--     | otherwise = find (p:ns) res parent
--     where
--         p = index parent n